src/tree/updater_gpu_hist.cu

```diff
@@ -626,6 +629,18 @@ struct GPUHistMakerDevice {
     return std::vector<DeviceSplitCandidate>(result_all.begin(), result_all.end());
   }
 
+  // Build gradient histograms for a given node across all the batches in the DMatrix.
+  void BuildHistBatches(int nidx, DMatrix* p_fmat) {
+    for (auto& batch : p_fmat->GetBatches<EllpackPage>(batch_param)) {
+      page = batch.Impl();
+      if (page->n_rows != row_partitioner->GetRows().size()) {
+        row_partitioner.reset();  // Release the device memory first before reallocating
+        row_partitioner.reset(new RowPartitioner(device_id, page->n_rows));
```

Comment:
I am wondering if it is better for us to keep a row_partitioner for each batch instead of resetting one of these objects. Memory usage will increase somewhat. Alternatively what is stopping us from creating the RowPartitioner object on the stack whenever we need it?

Reply:
I changed the `row_partitioner` back to using absolute row ids. PTAL